engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/MetadataStorage.java

```diff
@@ -0,0 +1,214 @@
+    */
+  public void update(ProcessingPass pass, ProcessingPass.Metadata newMeta) {
+    var copy = copyMetaMap();
+    copy.put(pass, newMeta);
+    metadata = copy;
+  }
+
+  /** Adds a metadata pair to the node metadata.
+    *
+    * This will overwrite any entry whose key matches [[MetadataPair#pass]].
+    *
+    * @param <K> the concrete type of the pass
+    * @param metadataPair the pair to add to the storage
+    */
+  public <K extends ProcessingPass> void update(MetadataPair<K> metadataPair) {
+    update(metadataPair.pass(), metadataPair.metadata());
+  }
+
+  /** Removes the metadata for the specified pass from the list.
+    *
+    * @param pass the pass to remove metadata for
+    * @tparam K the concrete type of `pass`
+    * @return the removed metadata for that pass, if it exists
+    */
+   public Option<ProcessingPass.Metadata> remove(ProcessingPass pass) {
+    var prev = metadata.get(pass);
+    if (prev == null) {
+      return Option.empty();
+    } else {
+      var copy = copyMetaMap();
+      copy.remove(pass);
+      metadata = copy;
+      return Option.apply(prev);
+    }
+   }
+
+  /** Gets the metadata for the specified pass.
+    *
+    * @param pass the pass to get the metadata for
+    * @tparam K the concrete type of `pass`
+    * @return the metadata for `pass`, if it exists
+    */
+  public Option<ProcessingPass.Metadata> get(ProcessingPass pass) {
+    var prev = (ProcessingPass.Metadata) metadata.get(pass);
+    return Option.apply(prev);
+  }
+
+  /** Creates a deep copy of `this`.
+    *
+    * @return a deep copy of `this`
+    */
+  public MetadataStorage duplicate() {
+    var map = new HashMap<ProcessingPass, ProcessingPass.Metadata>();
+    for (var entry : this.metadata.entrySet()) {
+      var key = entry.getKey();
+      var meta = (ProcessingPass.Metadata) entry.getValue();
+      var duplicated = meta.duplicate();
+      if (duplicated.nonEmpty()) {
+        map.put(key, duplicated.get());
+      }
+    }
+    var res = new MetadataStorage(map);
+    return res;
+  }
+
+
+  /** Maps across the stored metadata, transforming it to an output list.
+    *
+    * @param <R> the resulting element of the list
+    * @param fn the function to apply over the metadata
+    * @return a list containing the results of transforming the metadata storage
+    */
+  public <R> List<R> map(BiFunction<ProcessingPass, ProcessingPass.Metadata, R> fn) {
+    return metadata.entrySet().stream().map(
+      (en) -> fn.apply(en.getKey(), en.getValue())
+    ).toList();
+  }
+
+  /** Prepares the metadata for serialization.
+    *
+    * This operation takes place _in place_.
+    *
+    * Metadata prepared for serialization should not contain any links that
+    * span more than one module, or any other properties that are problematic
+    * when serialized.
+    *
+    * Due to the type safety properties of
+    * [[org.enso.compiler.core.ir.MetadataStorage]], to allow this conversion
```

Comment:
minor: convert scala doc to javadoc - replace stuff like `[[` with `{@code`, etc.